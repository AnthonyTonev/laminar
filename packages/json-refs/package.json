{
  "name": "@ovotech/json-refs",
  "version": "0.3.2",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "description": "Json Refs Resolver",
  "author": "Ivan Kerin <ikerin@gmail.com>",
  "repository": "git@github.com:ovotech/laminar.git",
  "license": "Apache-2.0",
  "devDependencies": {
    "@types/jest": "^25.2.3",
    "@types/nock": "^11.1.0",
    "@types/node": "^14.0.6",
    "@types/node-fetch": "^2.5.7",
    "@typescript-eslint/eslint-plugin": "^3.0.2",
    "@typescript-eslint/parser": "^3.0.2",
    "eslint": "^7.1.0",
    "eslint-config-prettier": "^6.11.0",
    "jest": "^26.0.1",
    "nock": "^12.0.3",
    "prettier": "^2.0.5",
    "ts-jest": "^26.0.0",
    "typescript": "^3.9.3"
  },
  "dependencies": {
    "js-yaml": "^3.14.0",
    "node-fetch": "^2.6.0"
  },
  "scripts": {
    "build": "tsc --declaration",
    "test": "jest test --runInBand",
    "lint:prettier": "prettier --list-different '{src,test}/**/*.ts'",
    "lint:eslint": "eslint '{src,test}/**/*.ts'",
    "lint": "yarn lint:prettier && yarn lint:eslint"
  },
  "jest": {
    "preset": "../../jest.json"
  }
}
