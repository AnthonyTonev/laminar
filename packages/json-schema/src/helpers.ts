import { Invalid, JsonSchema, Messages, Result, Schema, ValidateOptions, Validator } from './types';

export const childOptions = (name: string | number, options: ValidateOptions) => ({
  ...options,
  name: `${options.name}.${name}`,
});

export const isJsonSchema = (schema: Schema): schema is JsonSchema =>
  schema && typeof schema === 'object';

export const isObject = (value: any): value is { [key: string]: any } =>
  value && typeof value === 'object' && !Array.isArray(value);

export const isEqual = (a: any, b: any): boolean => {
  if (a === b) {
    return true;
  }
  if (a === undefined || b === undefined || a === null || b === null) {
    return false;
  }
  if (typeof a === 'object' && typeof b === 'object') {
    if (a instanceof Date && b instanceof Date) {
      return a.getTime() === b.getTime();
    } else if (Array.isArray(a) && Array.isArray(b)) {
      return a.length === b.length && a.every((item, index) => isEqual(item, b[index]));
    } else if (!Array.isArray(a) && !Array.isArray(b)) {
      const aKeys = Object.keys(a).sort();
      const bKeys = Object.keys(a).sort();
      return isEqual(aKeys, bKeys) && aKeys.every(key => isEqual(a[key], b[key]));
    }
  }
  return false;
};

export const flatten = <T>(arr: T[][]): T[] => {
  const result: T[] = [];
  for (const item of arr) {
    for (const inner of item) {
      result.push(inner);
    }
  }
  return result;
};

export const isUniqueWith = <T = any>(compare: (a: T, b: T) => boolean, array: T[]) => {
  const items: T[] = [];
  for (const item of array) {
    if (!items.some(prev => compare(prev, item))) {
      items.push(item);
    }
  }
  return items;
};

export const NoErrors = { errors: [] };
export const HasError = (code: keyof Messages, name: string, param?: any) => ({
  errors: [{ code, name, param }],
});
export const HasErrors = (errors: Invalid[]) => ({ errors });
export const CombineResults = (results: Result[]) =>
  HasErrors(flatten(results.map(result => result.errors)));

export const validateSchema: Validator<Schema> = (schema, value, options) => {
  if (schema === true) {
    return { errors: [], valid: true };
  } else if (schema === false) {
    return { errors: [{ code: 'false', name: options.name, param: false }], valid: false };
  } else if (schema) {
    return CombineResults(options.validators.map(validator => validator(schema, value, options)));
  } else {
    return NoErrors;
  }
};
