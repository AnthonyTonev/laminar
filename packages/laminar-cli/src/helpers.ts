import { compile, ensureValid, Schema } from '@ovotech/json-schema';
import { openapiV3 } from 'openapi-schemas';
import { OpenAPIObject } from 'openapi3-ts';
import { AstContext } from './traverse';
import * as YAML from 'yaml';

export const toString = async (stream: NodeJS.ReadStream): Promise<string> => {
  let str = '';

  return new Promise((resolve, reject) => {
    stream.on('data', (data) => (str += data.toString()));
    stream.on('end', () => resolve(str));
    stream.on('error', (error) => reject(error));
  });
};

export const toContext = async (
  fileName: string,
): Promise<{ context: AstContext; uris: string[]; value: OpenAPIObject }> => {
  const { schema, uris, refs } = await compile(fileName);
  const { value } = await ensureValid<OpenAPIObject>({
    schema: openapiV3 as Schema,
    value: schema,
    name: 'OpenAPI',
  });

  return { context: { root: value, refs }, value, uris };
};

export const parseSchema = (type: string, content: string): Schema => {
  switch (type) {
    case 'json':
      return JSON.parse(content);
    case 'yaml':
      return YAML.parse(content);
    default:
      throw new Error(`Unknown STDIN type: ${type}, accepts only "json" and "yaml"`);
  }
};

/**
 * regex for matching ANSI escape codes.
 *
 * Example:
 *
 * ```js
 *  '\u001B[4mcake\u001B[0m'.match(ansiRegex); => ['\u001B[4m', '\u001B[0m']
 * ```
 * Source: https://github.com/chalk/ansi-regex#readme
 */
export const ansiRegex = new RegExp(
  [
    '[\\u001B\\u009B][[\\]()#;?]*(?:(?:(?:[a-zA-Z\\d]*(?:;[-a-zA-Z\\d\\/#&.:=?%@~_]*)*)?\\u0007)',
    '(?:(?:\\d{1,4}(?:;\\d{0,4})*)?[\\dA-PR-TZcf-ntqry=><~]))',
  ].join('|'),
  'g',
);
