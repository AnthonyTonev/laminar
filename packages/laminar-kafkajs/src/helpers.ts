import type { ConfluentSchema } from '@kafkajs/confluent-schema-registry/dist/@types';
import type { logCreator, RecordMetadata } from 'kafkajs';
import { LoggerLike } from '@ovotech/laminar';
import { KafkaProducerService } from './kafka-producer.service';
import { EncodedMessage } from './types';

const stringLevel = { [0]: 'error', [1]: 'error', [2]: 'warn', [4]: 'info', [5]: 'debug' } as const;

/**
 * Convert a {@link LoggerLike} logger into a kafkajs logger.
 */
export function kafkaLogCreator(logger: LoggerLike): logCreator {
  return () => ({ level, log: { message, ...extra } }) => {
    logger[stringLevel[level]](message, extra);
  };
}

/**
 * A producer function, that has topic and schema baked in
 * Only needing the contents of the messages and the a {@link KafkaProducerService} object.
 *
 * @typeParam TValue The type of the the kafka message value, before its encoded
 */
export type Produce<TValue> = (
  producer: KafkaProducerService,
  messages: EncodedMessage<TValue>[],
) => Promise<RecordMetadata[]>;

/**
 * A helper to create a producer function, with "baked in" topic and schema names
 *
 * @typeParam TValue The type of the the kafka message value, before its encoded
 */
export function produce<TValue>(config: { topic: string; schema: ConfluentSchema }): Produce<TValue> {
  return (producer, messages) => producer.sendWithSchema<TValue>({ ...config, messages });
}
