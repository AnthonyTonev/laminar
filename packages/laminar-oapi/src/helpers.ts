const escapeRegExp = (str: string): string => str.replace(/[-\/\\^$*+?.()|[\]{}]/g, '\\$&');

export const toMatchPattern = (match: string): string =>
  match
    .split('/')
    .map((part) => (part === '*' ? '[^\\/]*' : escapeRegExp(part)))
    .join('\\/');

export const title = (str: string): string => str.replace(/^./, (first) => first.toUpperCase());
