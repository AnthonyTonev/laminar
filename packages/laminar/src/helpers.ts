export const toArray = (value: unknown): unknown[] =>
  Array.isArray(value) ? value : value ? [value] : [];

type Obj = Record<string, unknown>;

const isObj = (obj: unknown): obj is Obj => typeof obj === 'object' && obj !== null;

const setQuery = (path: string[], value: unknown, obj: Obj): Obj | unknown[] => {
  const [current, ...rest] = path;
  if (current) {
    const currentValue = obj[current];
    return {
      ...obj,
      [current]: rest.length
        ? setQuery(rest, value, isObj(currentValue) ? currentValue : {})
        : value,
    };
  } else {
    return toArray(rest.length ? setQuery(rest, value, obj) : value);
  }
};

const toQueryPath = (key: string): string[] => key.replace(/\]/g, '').split('[');

export const parseQueryObjects = (plain: Obj): Obj =>
  Object.entries(plain)
    .map<[string, unknown]>(([key, val]) => [
      key,
      typeof val === 'string' && val.includes(',') ? val.split(',') : val,
    ])
    .reduce((all, [key, val]) => setQuery(toQueryPath(key), val, all), {});
