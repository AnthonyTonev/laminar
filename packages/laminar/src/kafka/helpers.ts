import type { ConfluentSchema } from '@kafkajs/confluent-schema-registry/dist/@types';
import type { logCreator, RecordMetadata } from 'kafkajs';
import { LoggerLike } from '../logger';
import { KafkaProducerService } from './kafka-producer.service';
import { EncodedMessage } from './types';

const stringLevel = { [0]: 'error', [1]: 'error', [2]: 'warn', [4]: 'info', [5]: 'debug' } as const;

/**
 * Convert a {@link LoggerLike} logger into a kafkajs logger.
 */
export const kafkaLogCreator = (logger: LoggerLike): logCreator => () => ({ level, log: { message, ...extra } }) => {
  logger[stringLevel[level]](message, extra);
};

/**
 * A producer function, that has topic and schema baked in
 * Only needing the contents of the messages and the a {@link KafkaProducerService} object.
 */
export type Produce<TValue> = (
  producer: KafkaProducerService,
  messages: EncodedMessage<TValue>[],
) => Promise<RecordMetadata[]>;

/**
 * A helper to create a producer function, with "baked in" topic and schema names
 */
export const produce = <TValue>(config: { topic: string; schema: ConfluentSchema }): Produce<TValue> => (
  producer,
  messages,
) => producer.sendWithSchema<TValue>({ ...config, messages });
