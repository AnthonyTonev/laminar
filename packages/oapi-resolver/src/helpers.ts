const escapeRegExp = (str: string) => str.replace(/[-\/\\^$*+?.()|[\]{}]/g, '\\$&');

export const toMatchPattern = (match: string) =>
  match
    .split('/')
    .map(part => (part === '*' ? '[^\\/]*' : escapeRegExp(part)))
    .join('\\/');
