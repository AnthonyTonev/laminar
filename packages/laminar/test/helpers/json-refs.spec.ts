import { readFileSync } from 'fs';
import { join } from 'path';
import { resolveRefs } from '../../src/helpers/json-refs';

const openApi3 = JSON.parse(String(readFileSync(join(__dirname, 'openapi.v3.json'))));

describe('json-refs', () => {
  it('simple', async () => {
    const test1 = {
      test: '123',
      other: {
        $ref: '#/test',
      },
    };
    expect(resolveRefs(test1)).toMatchSnapshot();
  });

  it('openapi', async () => {
    expect(resolveRefs(openApi3)).toMatchSnapshot();
  });
});
