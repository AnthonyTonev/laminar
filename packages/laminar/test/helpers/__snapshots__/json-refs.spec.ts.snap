// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`json-refs openapi 1`] = `Promise {}`;

exports[`json-refs simple 1`] = `Promise {}`;
